# <div style="padding: 30px;color:white;margin:10;font-size:60%;text-align:left;display:fill;border-radius:10px;background-color:#FFFFFF;overflow:hidden;background-color:#15C3BA"><b><span style='color:#FFFFFF'>(12/14) |</span></b> <b>ПРАКТИКА</b></div>

### <b>Задание <span style='color:#15C3BA'>2.12.1</span></b> 
    
С помощью какого метода можно привести символы к нижнему регистру?
    
lower()
    
### <b>Задание <span style='color:#15C3BA'>2.12.2</span></b>
    
С помощью какого метода можно рассчитать расстояние между двумя точками?
    
Haversine 
    
### <b>Задание <span style='color:#15C3BA'>2.12.3</span></b>
    
Какими способами можно увеличить количество признаков в случае полиномиального создания признаков?
    
- Возведение в степень
- Перемножение
 
### <b>Задание <span style='color:#15C3BA'>2.12.4</span></b>
    
Какой атрибут (attribute) можно использовать у временного признака для применения к нему функций и взятия атрибутов?
    
dt
    
### <b>Задание <span style='color:#15C3BA'>2.12.5</span></b>
    
Какие можно сделать статистики по набору последовательных признаков? Например, средняя годовая зарплата за последние три года.
    
- Разность между годовыми зарплатами
- Минимальная средняя годовая зарплата
- Медиана средней годовой зарплаты
    
### <b>Задание <span style='color:#15C3BA'>2.12.6</span></b>
    
- Загрузите данные train.csv, оставьте в нем только 2 признака: **balance_due** и **payment_amount** (признаки должны следовать именно в указанном порядке). 
- Создайте из полученных признаков **полиномиальные признаки** степени 3. 
- Посчитайте **среднее значение для каждого получившегося признака** 
- В качестве ответа укажите индекс признака, который содержит максимальное среднее значение.


In [1]:
import numpy as np
import pandas as pd
%matplotlib inline

vis_data = pd.read_csv("train.csv", 
                       encoding = 'ISO-8859-1', 
                       low_memory = False)

# Выбираем только указаные признаки
ldf = vis_data[['balance_due','payment_amount']]
ldf

balance_due  payment_amount
0             305.0             0.0
1              75.0           780.0
2               0.0             0.0
3               0.0             0.0
4               0.0             0.0
...             ...             ...
250301          0.0             0.0
250302          0.0             0.0
250303          0.0             0.0
250304          0.0             0.0
250305          0.0             0.0

[250306 rows x 2 columns]

In [2]:
from sklearn.preprocessing import PolynomialFeatures

# Было 2 признака, стало 10 
model = PolynomialFeatures(3)
ldf_poly = model.fit_transform(ldf)
print(ldf.shape)
print(ldf_poly.shape)  

(250306, 2)
(250306, 10)


In [3]:
# Находим среднее число для каждого признака
array = np.mean(ldf_poly,axis=0)
print(array.argmax())

6


### <b>Задание <span style='color:#15C3BA'>2.12.7</span></b>

- Загрузите данные train.csv, найдите признак, обозначающий `дату`, когда был выписан штраф. 
- Найдите, сколько раз **штраф был выписан на выходных** и запишите это число в качестве ответа.

In [4]:
import numpy as np
import pandas as pd
%matplotlib inline

vis_data = pd.read_csv("train.csv", 
                       encoding = 'ISO-8859-1', 
                       low_memory = False)

vis_data.columns

Index(['ticket_id', 'agency_name', 'inspector_name', 'violator_name',
       'violation_street_number', 'violation_street_name',
       'violation_zip_code', 'mailing_address_str_number',
       'mailing_address_str_name', 'city', 'state', 'zip_code',
       'non_us_str_code', 'country', 'ticket_issued_date', 'hearing_date',
       'violation_code', 'violation_description', 'disposition', 'fine_amount',
       'admin_fee', 'state_fee', 'late_fee', 'discount_amount',
       'clean_up_cost', 'judgment_amount', 'payment_amount', 'balance_due',
       'payment_date', 'payment_status', 'collection_status',
       'grafitti_status', 'compliance_detail', 'compliance'],
      dtype='object')

In [5]:
# У этого фичера dtype = object
vis_data['ticket_issued_date']

0         2004-03-16 11:40:00
1         2004-04-23 12:30:00
2         2004-04-26 13:40:00
3         2004-04-26 13:30:00
4         2004-04-26 13:00:00
                 ...         
250301    2010-12-02 11:00:00
250302    2010-12-02 11:00:00
250303    2010-12-02 11:00:00
250304    2010-12-02 11:00:00
250305    2010-12-02 11:00:00
Name: ticket_issued_date, Length: 250306, dtype: object

In [6]:
issued_date = vis_data['ticket_issued_date']
issued_date = issued_date.astype('datetime64[ns]')
issued_date

0        2004-03-16 11:40:00
1        2004-04-23 12:30:00
2        2004-04-26 13:40:00
3        2004-04-26 13:30:00
4        2004-04-26 13:00:00
                 ...        
250301   2010-12-02 11:00:00
250302   2010-12-02 11:00:00
250303   2010-12-02 11:00:00
250304   2010-12-02 11:00:00
250305   2010-12-02 11:00:00
Name: ticket_issued_date, Length: 250306, dtype: datetime64[ns]

In [7]:
weekend_index = issued_date[issued_date.dt.dayofweek > 4].index
print(weekend_index.shape)

weekday_index = issued_date[issued_date.dt.dayofweek < 5].index
print(weekday_index.shape)

(1620,)
(248686,)


- 1620 штрафов было выписано на выходных и 248686 в будние дни

<br>    
    
## Практика

**Перекодировать признаки**

- Ранее мы уже выяснили, что данные о квартирах не все числовые. 
- Есть и признаки, которые содержат текстовые значения : `ординальный` (уровень экологии) и `категориальный` (район). 
- Нам обязательно нужно **перекодировать эти признаки**, так как иначе мы не сможем в дальнейшем работать с этим датасетом и применять на нем какие-то математические алгоритмы

**Ординальные признаки**

- мы можем перекодировать в числовой довольно простым способом: просто заменить каждое значение на число (4 - 'excellent', 3 - 'good' etc.).

Для этого необходимо:

 создать словарь, , сопоставляющий числовые значения от 0 до 4 описанию экологии от 'no data' до 'excellent' соответственно
 заменить значения в колонке ecology на заданные нами выше с помощью метода .replace :

In [8]:
import pandas as pd

df = pd.read_csv('data_flats.csv',delimiter=';')
suburb_dict = dict(tuple(df.groupby('sub_area')))
df.columns

Index(['id', 'full_sq', 'life_sq', 'floor', 'sub_area', 'preschool_quota',
       'preschool_education_centers_raion', 'school_quota',
       'school_education_centers_raion',
       'school_education_centers_top_20_raion', 'hospital_beds_raion',
       'healthcare_centers_raion', 'university_top_20_raion',
       'sport_objects_raion', 'additional_education_raion',
       'culture_objects_top_25_raion', 'shopping_centers_raion',
       'office_raion', 'metro_min_avto', 'metro_km_avto', 'metro_min_walk',
       'metro_km_walk', 'kindergarten_km', 'school_km', 'park_km',
       'green_zone_km', 'industrial_km', 'railroad_station_walk_km',
       'railroad_station_walk_min', 'public_transport_station_km',
       'public_transport_station_min_walk', 'mkad_km', 'ttk_km', 'sadovoe_km',
       'bulvar_ring_km', 'kremlin_km', 'big_market_km', 'market_shop_km',
       'fitness_km', 'swim_pool_km', 'ice_rink_km', 'stadium_km',
       'basketball_km', 'hospice_morgue_km', 'university_km', 'workp

In [9]:
df['ecology'].value_counts()

poor            8017
no data         7656
good            7173
excellent       3938
satisfactory    3685
Name: ecology, dtype: int64

In [10]:
ecology_dict = {'excellent':4,'good':3,'satisfactory':2,'poor':1,'no data':0}
df.ecology = df.ecology.replace(to_replace=ecology_dict)

In [11]:
df['ecology']

0        3
1        4
2        1
3        3
4        4
        ..
30464    3
30465    1
30466    0
30467    2
30468    1
Name: ecology, Length: 30469, dtype: int64

**Район** (sub_area)

- Теперь необходимо разобраться с переменной, содержащей информацию о районе.  
- Для этого воспользуйтесь методом `get_dummies`

### <b>Задание <span style='color:#15C3BA'>2.12.8</span></b>

Сколько теперь признаков в датасете после перекодировки?

In [12]:
dummy_features = pd.get_dummies(df['sub_area']).shape[1]

print(f'total number of features: {df.shape[1] + dummy_features}')

total number of features: 202


Найдите среднее арифметическое для признака "Экология". Округлите до сотых

In [13]:
df['ecology'].mean()

1.7282483836030063